src/Functions/parseDateTime.cpp

```diff
@@ -0,0 +1,1747 @@
+    constexpr Int32 cumulativeYearDays[]
+        = {0,     365,   730,   1096,  1461,  1826,  2191,  2557,  2922,  3287,  3652,  4018,  4383,  4748,  5113,  5479,  5844,  6209,
+           6574,  6940,  7305,  7670,  8035,  8401,  8766,  9131,  9496,  9862,  10227, 10592, 10957, 11323, 11688, 12053, 12418, 12784,
+           13149, 13514, 13879, 14245, 14610, 14975, 15340, 15706, 16071, 16436, 16801, 17167, 17532, 17897, 18262, 18628, 18993, 19358,
+           19723, 20089, 20454, 20819, 21184, 21550, 21915, 22280, 22645, 23011, 23376, 23741, 24106, 24472, 24837, 25202, 25567, 25933,
+           26298, 26663, 27028, 27394, 27759, 28124, 28489, 28855, 29220, 29585, 29950, 30316, 30681, 31046, 31411, 31777, 32142, 32507,
+           32872, 33238, 33603, 33968, 34333, 34699, 35064, 35429, 35794, 36160, 36525, 36890, 37255, 37621, 37986, 38351, 38716, 39082,
+           39447, 39812, 40177, 40543, 40908, 41273, 41638, 42004, 42369, 42734, 43099, 43465, 43830, 44195, 44560, 44926, 45291, 45656,
+           46021, 46387, 46752, 47117, 47482, 47847, 48212, 48577, 48942, 49308, 49673};
+
+    Int64 numLiteralChars(const char * cur, const char * end)
+    {
+        bool found = false;
+        Int64 count = 0;
+        while (cur < end)
+        {
+            if (*cur == '\'')
+            {
+                if (cur + 1 < end && *(cur + 1) == '\'')
+                {
+                    count += 2;
+                    cur += 2;
+                }
+                else
+                {
+                    found = true;
+                    break;
+                }
+            }
+            else
+            {
+                ++count;
+                ++cur;
+            }
+        }
+        return found ? count : -1;
+    }
+
+    struct Date
+    {
+        Int32 year = 1970;
+        Int32 month = 1;
+        Int32 day = 1;
+        std::vector<Int32> day_of_month_values;
+        bool is_ad = true; // AD -> true, BC -> false.
+
+        Int32 week = 1; // Week of year based on ISO week date, e.g: 27
+        Int32 day_of_week = 1; // Day of week, Monday:1, Tuesday:2, ..., Sunday:7
+        bool week_date_format = false;
+
+        Int32 day_of_year = 1;
+        std::vector<Int32> day_of_year_values;
+        bool day_of_year_format = false;
+
+        bool century_format = false;
+
+        bool is_year_of_era = false; // Year of era cannot be zero or negative.
+        bool has_year = false; // Whether year was explicitly specified.
+
+        Int32 hour = 0;
+        Int32 minute = 0;
+        Int32 second = 0;
+
+        bool is_am = true; // AM -> true, PM -> false
+        bool is_clock_hour = false; // Whether most recent hour specifier is clockhour
+        bool is_hour_of_half_day = false; // Whether most recent hour specifier is of half day.
+
+        std::optional<Int64> time_zone_offset;
+
+        void reset()
+        {
+            year = 1970;
```

Comment:
There is a lot of state in the form of boolean state variables, creating complex interactions between them). Please document the meaning of each state variable, otherwise future readers will have a hard time.

(I guess a better alternative would be to omit all the boolean state variables and instead maintain the state implicitly by storing all numeric values in `std::optional`s., e.g. `std::optional<Int32> year`)

Reply:
I think using bool is more precise than std::optional. For example, year is valid when one of  week_date_format or day_of_year_format or week_date_format is true. Besides, too much std::optional will cause performance degradation for memory access in heap. 

I will add more docs for every field of struct DateTime. Thanks for reminding.

Reply:
Comments of every field of `DateTime` is added. BTW, I replace `std::optional<Int64> time_zone_offset` to two fields: `bool has_time_zone_offset` and `Int32 time_zone_offset` to make `DateTime` a POD struct for performance benefits.